src/libraries/Microsoft.Extensions.Logging.Abstractions/ref/Microsoft.Extensions.Logging.Abstractions.cs

```diff
@@ -133,11 +133,11 @@ namespace Microsoft.Extensions.Logging.Abstractions
         private readonly TState _State_k__BackingField;
         private readonly object _dummy;
         private readonly int _dummyPrimitive;
-        public LogEntry(Microsoft.Extensions.Logging.LogLevel logLevel, string category, Microsoft.Extensions.Logging.EventId eventId, TState state, System.Exception exception, System.Func<TState, System.Exception, string> formatter) { throw null; }
+        public LogEntry(Microsoft.Extensions.Logging.LogLevel logLevel, string category, Microsoft.Extensions.Logging.EventId eventId, TState state, System.Exception? exception, System.Func<TState, System.Exception?, string>? formatter) { throw null; }
         public string Category { get { throw null; } }
```

Comment:
Structs like this are tough.  While the constructor declaring `string category` as non-nullable means we wouldn't expect any nulls here, it is possible for someone to do `LogEntry<TState> le = default;` or `LogEntry<TState> le = new LogEntry<TState>()` (which are equivalent), in which case `le.Category` will be null.  So, this comes down to a judgement call as to how common it'll be for someone to work with LogEntry's, might they use them zero-initialized, etc.  If the answer is "No, the 99.9% case is a developer only uses instances produced by the framework, and those will always have been constructed correctly before being handed out", then the current annotation looks good.  If instead the answer is "Yes, there's a non-negligible chance a developer will find themselves working with a LogEntry that may have its default value", then Category and State should both be marked `?` as well.

Reply:
In the repo today, LogEntry is only used for console logging and specifically on the `ConsoleFormatter.Write<TState>` API which receives a LogEntry instance. As you also mentioned, it is expected that mostly the developer just uses the instance provided/instantiated by the framework on the first line below:

https://github.com/dotnet/runtime/blob/23611e16a93c9056ceeb27c55e93809e1baf56c4/src/libraries/Microsoft.Extensions.Logging.Console/src/ConsoleLogger.cs#L45-L46

and that gets passed onto a console formatter in the next line, which may or may not be provided by the framework.

In a situation where (in the future) this pattern gets adopted by other loggers (e.g. DebugLogger, and not just ConsoleLogger), then what would happen is that `ILogger.Log<TState>(logLevel, eventId, state, exception, formatter)` would be responsible for instantiating and passing down LogEntry to a `Formatter.Write<TState>(logEntry, ...)` API which ends up just using it and doing custom formatting.

To better understand whether or not LogEntry properties should be nullable and I looked at more usages of:
```c#
ILogger.Log<TState>(LogLevel logLevel, EventId eventId, TState state, Exception? exception, Func<TState, Exception?, string> formatter);
```
in the framework. 

- [x] **`TState state`** and **`Exception? exception`**: up to caller to pass null. OK as nullable.
- [x] **`LogLevel logLevel`** is an enum
- [x] **`EventId eventId`**  is a struct
- [x] **`string category`**: I believe category should not be nullable. 

e.g for ConsoleLogger
https://github.com/dotnet/runtime/blob/5e5c4ebf1df276103b5a923a14325f845c2e7177/src/libraries/Microsoft.Extensions.Logging.Console/src/ConsoleLogger.cs#L45
`_name` cannot be null.

Also checked LoggerFactory in Microsoft.Extensions.Logging which helps setup different LoggerProviders (ConsoleLogger being one of them), looking at its `CreateLogger(string categoryName)` API below, confirmed it would also throw ArgumentNullException when null category passed in.
so something like 
```c#
public static async Task Main(string[] args)
        {
            using var loggerFactory = LoggerFactory.Create(builder => { builder.AddConsole();  });

            var logger = loggerFactory.CreateLogger(categoryName); // when null throws ArgumentNullException
            logger.Log(LogLevel.Information, 0, new Exception(), null, (t, s) => string.Empty);
        }
```

- [ ] **`Func<TState, Exception?, string> formatter`**: Did a similar investigation for the formatter argument, and I believe formatter should not be nullable either,
For EventLogLogger, ConsoleLogger and EventSourceLogger, passing null formatter throws exception. 

There is also a NullLogger that is a noop on Log, therefore not throwing ArgumentNullException. And TraceSourceLogger allows null formatter as seen below (handles it rather than throwing):

https://github.com/dotnet/runtime/blob/5cb5ae9f33837ec63e1a444f5d8a3f64d2d7188b/src/libraries/Microsoft.Extensions.Logging.TraceSource/src/TraceSourceLogger.cs#L28-L35

Would it be reasonable to restrict `formatter` argument in `ILogger.Write<TState>(.., formatter)` to be not nullable and make TraceSourceLogger consistent with the others through this change? 

---

Back to your question on "how common it would be for someone to work with LogEntry's", I think it was just introduced on 5.0 to conveniently pass down ILogger.Log<TState> arguments onto Formatter.Write<TState> and for that it is not expected that someone would get use of default `LogEntry<TState> le = default;` or `LogEntry<TState> le = new LogEntry<TState>()`.

Reply:
Chatted offline, the conclusion is to have ILogger.Log's formatter argument to be non-nullable